Add UI for adding, renaming, and deleting a color scheme

carlos-zamora opened the pull request:
Introduces the following UI controls to the ColorSchemes page:
- "Add new" button
  - next to dropdown selector
  - adds a new color scheme named ("Color Scheme #" where # is the number of color schemes you have)
- "Rename" Button
  - next to the selector
  - replaces the ComboBox with a TextBox and the accept/cancel buttons appear
- "Delete" button
  - bottom of the page
  - opens flyout, when confirmed, deletes the current color scheme and selects another one

This also adds a Delete button to the Profiles page. The Hide checkbox was moved above the Delete button.

## References
#1564 - Settings UI
#6800 - Settings UI Completion Epic

## Detailed Description of the Pull Request / Additional comments

**Color Schemes:**
- Deleting a color scheme selects another one from the list available
- Rename replaces the combobox with a textbox to allow editing
- The Add New button creates a new color scheme named "Color Scheme X" where X is the number of schemes defined
- In-box color schemes cannot be deleted

**Profile:**
- Deleting a profile selects another one from the list available
- the rename button does not exist (yet), because it needs a modification to the NavigationView's Header Template
- The delete button is disabled for in-box profiles (CMD and Windows Powershell) and dynamic profiles

## Validation Steps Performed
**Color Schemes - Add New**
✅ Creates a new color scheme named "Color Scheme X" (X being the number of color schemes)
✅ The new color scheme can be renamed/deleted/modified

**Color Schemes - Rename**
✅ You cannot rename an in-box color scheme
✅ The rename button has a tooltip
✅ Clicking the rename button replaces the combobox with a textbox
✅ Accept --> changes name
✅ Cancel --> does not change the name
✅ accepting/cancelling the rename operation updates the combo box appropriately

**Color Schemes - Delete**
✅ Clicking delete produces a flyout to confirm deletion
✅ Deleting a color scheme removes it from the list and select the one under it
✅ Deleting the last color scheme selects the last available color scheme after it's deleted
✅ In-box color schemes have the delete button disabled, and a disclaimer appears next to it

**Profile- Delete**
✅ Base layer presents a disclaimer at the top, and hides the delete button
✅ Dynamic and in-box profiles disable the delete button and show the appropriate disclaimer next to the disabled button
✅ Clicking delete produces a flyout to confirm deletion
✅ Regular profiles have a delete button that is styled appropriately
✅ Clicking the delete profile button opens a content dialog. Confirmation deletes the profile and navigates to the profile indexed under it (deleting the last one redirects to the last one)


## Demo
Refer to this post [here](https://github.com/microsoft/terminal/pull/8403#issuecomment-747545651.
Confirmation flyout demo: https://github.com/microsoft/terminal/pull/8403#issuecomment-747657842

carlos-zamora commented on src/cascadia/TerminalSettingsModel/ColorScheme.cpp:
```diff
@@ -46,15 +46,44 @@ ColorScheme::ColorScheme() :
+}
+
+// Method Description:
+// - Sets the color table to Cambell's values
```
```suggestion
// - Sets the color table to Campbell's values
```

DHowett commented:
1. Should we use the well-known + iconography for the Add button?
2. Should we use the well-known Trash Can iconography for the Delete button?
3. Should the delete button be destructive red?
4. Same question as #8154: should we use a confirmation flyout on Delete?
![image](https://user-images.githubusercontent.com/189190/100280471-16eb2b80-2f1d-11eb-9a98-72c16ba0c51c.png)

DHowett commented:
5. Should rename be a button that opens a little dialog that prompts for the name? How do other Windows applications that let you edit the names of things handle this?

DHowett commented:
This is a big destructive rename operation:

![image](https://user-images.githubusercontent.com/189190/100280631-59146d00-2f1d-11eb-8795-13516fce14f8.png)

We probably don't need something THAT big.

DHowett commented on src/cascadia/TerminalSettingsModel/ColorScheme.cpp:
```diff
@@ -46,15 +46,44 @@ ColorScheme::ColorScheme() :
+// - Sets the color table to Cambell's values
+void ColorScheme::_InitializeColorTable()
 {
+    _table.at(0) = Utils::ColorFromHexString("#0C0C0C");
+    _table.at(1) = Utils::ColorFromHexString("#C50F1F");
+    _table.at(2) = Utils::ColorFromHexString("#13A10E");
+    _table.at(3) = Utils::ColorFromHexString("#C19C00");
+    _table.at(4) = Utils::ColorFromHexString("#0037DA");
+    _table.at(5) = Utils::ColorFromHexString("#881798");
+    _table.at(6) = Utils::ColorFromHexString("#3A96DD");
+    _table.at(7) = Utils::ColorFromHexString("#CCCCCC");
+    _table.at(8) = Utils::ColorFromHexString("#767676");
+    _table.at(9) = Utils::ColorFromHexString("#E74856");
+    _table.at(10) = Utils::ColorFromHexString("#16C60C");
+    _table.at(11) = Utils::ColorFromHexString("#F9F1A5");
+    _table.at(12) = Utils::ColorFromHexString("#3B78FF");
+    _table.at(13) = Utils::ColorFromHexString("#B4009E");
+    _table.at(14) = Utils::ColorFromHexString("#61D6D6");
+    _table.at(15) = Utils::ColorFromHexString("#F2F2F2");
```
delete this pull request

DHowett commented on src/cascadia/TerminalSettingsModel/ColorScheme.idl:
```diff
@@ -4,14 +4,16 @@
         Windows.UI.Color CursorColor;
 
-        Windows.UI.Color[] Table { get; };
+        Windows.UI.Color[] Table();
```
note why with a comment.

DHowett reviewed (changes requested):
Yea

carlos-zamora commented:
Merging feedback from @DHowett (DH) and @htcfreek (HTC) here since there's some overlap.

> (**DH**) 1. Should we use the well-known + iconography for the Add button?
> (**DH**) 2. Should we use the well-known Trash Can iconography for the Delete button?

Added. Good call 😊. Updating the demo images with it.

> (**DH**) 3. Should the delete button be destructive red?
> (**HTC**) 1. I would paint the background of the delete buttons red.

I've been searching for this destructive red style for quite a bit now. But I can't find it! Does anybody know what it is? There has to be a built-in one in XAML right? I don't want to hand-roll it then have to hook it up a bunch like our MinMaxClose Control if I don't have to :(

> (**DH**) 4. Same question as #8154: should we use a confirmation flyout on Delete?
> (**DH**) 5. Should rename be a button that opens a little dialog that prompts for the name? How do other Windows applications that let you edit the names of things handle this?
> (**HTC**) 2. I would move the "delete profile" bottun to the general tab behind the name field.
> (**HTC**) 3. I would move the "delete color scheme" behind the name field.
> (**HTC**) 1. Showing the color shemes in the same way as the profiles in the nav pane would bring more consistency.
> (**HTC**) 2. Is it possible to replace the delete button with an trash icon on the rigth side of the nav item which occurs on hovering?

Summoning @cinnamon-msft to redesign some of this. I'm thinking that maybe we could do something more compact like this:
![Compact Mock Up](https://user-images.githubusercontent.com/11050425/100293285-0a280100-2f38-11eb-90c2-a22d8e69b48f.png)
We'd add tooltips on each button saying something like...
- Add a new color scheme
- Rename the color scheme
- Delete the color scheme

Delete is disabled for in-box color schemes (or the last color scheme)

Rename and Delete would open a dialog with a text box and confirmation statement respectively.

Looking to the future, at some point, we probably want a "duplicate color scheme" option. It'd be nice if we just came up with that design right now. The implementation wouldn't be difficult either. Just a matter of needing the design.

For my proposed version above, it would be expanded to look something like this:
![Expanded Compact Mock Up](https://user-images.githubusercontent.com/11050425/100293553-f16c1b00-2f38-11eb-9d2d-f6788da279c3.png)

DHowett commented:
> have to hook it up a bunch like our MinMaxClose Control

whoa boy hold up

that's in a totally different library serving a totally different purpose. the red and the hover-red in use there are _very specific to the windows platform's close button_. Perish the thought of deduplicating those.

carlos-zamora commented:
I'm also wondering if we could add these buttons to the Header at the top? For profiles, the header already updates with the profile's name. Maybe we should add a Rename and Delete button there?

For Color Schemes, maybe we _should_ be updating according to which color scheme you're looking at? Then follow the same model and add the rename/delete button there. But keep the add new and duplicate buttons next to the combo box?

Not sure. Open to discussion. Either way, I do agree that the "Name" text box looks a bit wonky.

carlos-zamora commented:
# New Design

@cinnamon-msft and I met up today to discuss and apply some of the feedback we got in this PR. Here's a summary of the results from that discussion.

## Color Schemes Page
Rough Design:
![Color Schemes Page Mockup](https://user-images.githubusercontent.com/11050425/100807384-105e2780-33e7-11eb-8215-9bcce6aa0ad7.png)

![Rename Dialog](https://user-images.githubusercontent.com/11050425/100807680-9b3f2200-33e7-11eb-80ef-cb90f6bfe462.png)


Notes:
- Rename Button
  - Placed next to dropdown (also removed the Name text box)
  - tooltip = "Rename"
  - On click, opens a dialog (_very_ rough design shown above). This is prepopulated with the current name, and allows you to save/cancel a rename operation. On save (primary button), changes are propagated.
- Add New Button
  - Placed next to rename button
  - Includes both icon _and_ "Add new" text
  - Just like today, it creates a new "Campbell" color scheme with a preset name, and navigates to it. (This behavior can/will probably change later? Especially when we have a discussion on the "Duplicate" button)
- Delete Button
  - Placed at the bottom of the page
  - Must adhere to XAML destructive styling (still searching for it)

Things to work on after this PR lands:
- Duplicate button (design needed)
- Prevent modifying in-box color schemes

## Profile Page
_Very_ Rough Design:
![Profile Page Mockup](https://user-images.githubusercontent.com/11050425/100808159-7a2b0100-33e8-11eb-99af-ffb879078897.png)

Notes:
- Rename Button
  - Placed in header, next to profile name (also removed the Name text box)
  - Follows consistent interaction story from Color Scheme renaming
- Delete Button
  - Follows consistent interaction story from Color Scheme deleting (in General page)

Things to work on after this PR lands:
- propagate rename/delete (and new icon) to NavigationView

DHowett commented:
pull with a merge -- the rebase will have fucked this right up

carlos-zamora commented:
The PR isn't ready yet. The buttons are styled now. And the Color Schemes page is basically done. Just missing...
- [ ] Rename button --> popup dialog
- [ ] rename Profile
- [ ] delete Profile

carlos-zamora commented:
## Demo

### Color Schemes
![General View](https://user-images.githubusercontent.com/11050425/102148933-3e4d5e00-3e22-11eb-8cfd-fb50a548ef45.png)
![In-Box Scheme](https://user-images.githubusercontent.com/11050425/102148942-43121200-3e22-11eb-9ad5-a6289a91b11f.png)
![Rename Dialog](https://user-images.githubusercontent.com/11050425/102148992-5de48680-3e22-11eb-87fd-63fa0919a28f.png)

### Profiles
![General View](https://user-images.githubusercontent.com/11050425/102149062-7bb1eb80-3e22-11eb-95f3-0553c71df226.png)
![In-Box Profile](https://user-images.githubusercontent.com/11050425/102149080-840a2680-3e22-11eb-80cc-70457c93552e.png)

DHowett commented on src/cascadia/TerminalSettingsEditor/ColorSchemes.cpp:
```diff
@@ -153,6 +177,24 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
+    void ColorSchemes::Rename_Click(IInspectable const& sender, RoutedEventArgs const& /*e*/)
+    {
+        const auto schemeName{ CurrentColorScheme().Name() };
+        const hstring title{ fmt::format(std::wstring_view{ RS_(L"ColorScheme_RenameDialog/Title") }, schemeName) };
+
+        auto dialog{ FindName(L"RenameDialog").try_as<ContentDialog>() };
+        dialog.Title(winrt::box_value(title));
```
this looks like an `x:Uid` exploded and left shrapnel everywhere. What happened?

DHowett replied:
ah, it's because you wanted to format the color scheme name into it. Sure, fine.

DHowett commented on src/cascadia/TerminalSettingsEditor/ColorSchemes.cpp:
```diff
@@ -153,6 +177,24 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
+        dialog.XamlRoot(sender.as<UIElement>().XamlRoot());
+
+        NameBox().Text(schemeName);
+        dialog.ShowAsync(ContentDialogPlacement::Popup);
+    }
+
+    void ColorSchemes::RenameConfirm_Click(IInspectable const& /*sender*/, ContentDialogButtonClickEventArgs const& /*e*/)
+    {
+        CurrentColorScheme().Name(NameBox().Text());
```
you can do this without an event handler on the button (and therefore the complexity) by `co_await`ing the `ShowAsync` and checking whether it was confirmed.

DHowett commented on src/cascadia/TerminalSettingsEditor/Profiles.xaml:
```diff
@@ -85,7 +85,10 @@ the MIT License. See LICENSE in the project root for license information. -->
+                            Style="{StaticResource BaseButtonStyle}"
+                            Margin="{StaticResource StandardControlSpacing}"
+                            IsEnabled="{x:Bind State.CanDeleteProfile}">
                         <Button.Resources>
```
got a flyout?

DHowett reviewed (commented):
As a person who enters a blind rage (I kid) when he encounters a button that is disabled for what appears to be _no reason whatsoever_, I would very much like for us to have a tooltip or a disclaimer textblock in a slightly smaller font size explaining why a profile or colorscheme cannot be deleted.

DHowett commented:
You also cannot rename an inbox scheme right? it'll just hydra itself and create another copy?

mdtauk commented:
Is there a need to add a preview onto the colour scheme page, showing sample text on the background colour - to visualise how the colours look together?

ConHost has one...
![image](https://user-images.githubusercontent.com/7389110/102149601-c190a300-3e66-11eb-9a59-b2244160d98f.png)

carlos-zamora commented:
@DHowett (or to anybody else that comes along), there's a few bugs/questions I've hit but I'm not sure what to do about them:
- Rename Dialog in Dark Mode looks broken:
![Dark Mode - Rename Dialog](https://user-images.githubusercontent.com/11050425/102149543-933da400-3e23-11eb-9ae8-a1dab3711f43.png)
- The Rename Dialog does _not_ accept text input. But I've done nothing differently with it.
- Should the disabled delete button still have a red tint?
- Should we disable deletion of dynamic profiles? (at least until we have a way to retain state)

mdtauk commented:
> * Rename Dialog in Dark Mode looks broken:
>   ![Dark Mode - Rename Dialog](https://user-images.githubusercontent.com/11050425/102149543-933da400-3e23-11eb-9ae8-a1dab3711f43.png)
>
It looks like the "smoke" colour for the ContentDialog is still using the Light Theme resource, not sure if its a Terminal implementation problem, or a problem with the ContentDialog templates/styles.


> * Should the disabled delete button still have a red tint?
>
There is a ThemeResource called System Error Text.  In the Light theme it appears red, but in dark theme, it is a yellow colour.  Perhaps you could use this resource?  The InfoBar has Theme Resources for its **Severity** colours,  Perhaps you could use the critical colour from this?  You may want to confirm with the design teams

> * Should we disable deletion of dynamic profiles? (at least until we have a way to retain state)
>

Is there a way to add an option for profiles to hide/disable them rather than removing them?  So they do not show up in the New Tab dropdown, but still show up in Settings?

DHowett commented:
For dynamic profiles, we should consider replacing delete with hide -- same trash can icon, however. How should we designate hidden profiles in the navigation view?

mdtauk commented:
> For dynamic profiles, we should consider replacing delete with hide -- same trash can icon, however. How should we designate hidden profiles in the navigation view?

Perhaps append _(hidden)_ to the items, or maybe adding a hidden icon?

![image](https://user-images.githubusercontent.com/7389110/102150894-85ab0d00-3e69-11eb-883d-0435ee62e1ee.png)

carlos-zamora commented:
> Is there a need to add a preview onto the colour scheme page, showing sample text on the background colour - to visualise how the colours look together?

I really like this idea. Throwing it into #6800

mdtauk commented:
> > Is there a need to add a preview onto the colour scheme page, showing sample text on the background colour - to visualise how the colours look together?
> 
> I really like this idea. Throwing it into #6800

Not sure what kind of sample text could  demonstrate all the colour combinations, but there is the Color Tool
![image](https://user-images.githubusercontent.com/7389110/102154640-bb53f400-3e71-11eb-8c52-ccf4e0769573.png)

zadjii-msft commented:
We've had a ton of problems with content dialogs in the past. We've got all sorts of hacks to apply the app theme to the dialogs, because they're technically not parented to the `App` (or something crazy like that). So I wouldn't worry _too_ much about that.

The Dialog not receiving input though - that's pretty bad. My gut is telling me that this is a XAML Islands bug, but I can't be sure. 

IMO the disabled "delete \<foo>" button being gray makes sense to me. That's pretty clearly "not actionable", especially when compared with the red one.

If the "delete" button is wired up to "hide" dynamic profiles, then let's do it. Otherwise, won't they just re-appear in the UI after saving? If that's the case, definitely disable the button.

zadjii-msft commented on src/cascadia/TerminalSettingsModel/TerminalSettingsSerializationHelpers.h:
```diff
@@ -168,7 +167,7 @@ struct ::Microsoft::Terminal::Settings::Model::JsonUtils::ConversionTrait<::winr
         }
 
         ::winrt::Windows::UI::Text::FontWeight weight{
-            static_cast<uint16_t>(std::clamp(value, 100u, 990u))
```
Wait is this one intentional?

carlos-zamora replied:
whoops! bad merge! Good catch

zadjii-msft commented on src/cascadia/TerminalSettingsModel/ColorScheme.cpp:
```diff
@@ -46,15 +47,22 @@ ColorScheme::ColorScheme() :
-    _Background{ defaultBg },
-    _SelectionBackground{ DEFAULT_FOREGROUND },
-    _CursorColor{ cursorColor }
+    ColorScheme{ name }
```
Wait couldn't this still be an initializer list?
```suggestion
    ColorScheme{ name },
	_Foreground{ defaultFg },
    _Background{ defaultBg },
    _SelectionBackground{ DEFAULT_FOREGROUND },
    _CursorColor{ cursorColor }
```

my c++fu is not strong, but I thought this worked

carlos-zamora replied:
> a delegating constructor cannot have other mem-initializers

😢

DHowett replied:
The correct way for you to do this is to delegate everything to the _largest_ constructor.

zadjii-msft commented on src/cascadia/TerminalSettingsEditor/MainPage.cpp:
```diff
@@ -230,6 +231,23 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
+
+        // You cannot delete the Windows Powershell and CMD profiles
+        const auto myGuid{ profile.Guid() };
+        const winrt::guid pwshGuid{ ::Microsoft::Console::Utils::GuidFromString(L"{61c54bbd-c2c6-5271-96e7-009a87ff44bf}") };
```
These two can probably be `const static`, so they're only ever initialized _once_.

zadjii-msft commented on src/cascadia/TerminalSettingsEditor/MainPage.cpp:
```diff
@@ -230,6 +231,23 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
+        if (myGuid != pwshGuid && myGuid != cmdGuid)
+        {
+            auto pfnDeleteProfile{ std::bind(&MainPage::_DeleteProfile, this, profile) };
+            state->SetDeleteProfileCallback(pfnDeleteProfile);
```
Okay so this whole thing is _weird_. I get it, but I think it needs some comments. Lemme make sure I followed this right:
* The MainPage has access to the whole settings model, but the profile page only has access to the Profile. So if the profile page wants to be able to delete a profile, it needs to ask the MainPage to do it.
* Not all profiles are deletable (namely, `cmd` and `powershell`).
* We're passing in a callback to the profile page if the profile is deletable.
* When the delete button is pressed, we'll call that callback, and MainPage will delete the profile.

<hr>

1. Is there any reason we didn't just go with a normal event for this? Something like `Profiles::DeleteRequested`, that `MainPage` listens for? 
2. Should the availability of the delete button be in the Profile ViewModel? Seems weird to make the MainPage responsible for it here. Esp. considering the future where dynamic profiles (and defaults) might need to be "hidden", not "deleted"

carlos-zamora replied:
Your explanation is spot on. I'll add a comment documenting the whole thing.

> 1. Is there any reason we didn't just go with a normal event for this? Something like `Profiles::DeleteRequested`, that `MainPage` listens for? 

No particular reason. I'm not familiar with the benefits/tradeoffs of each approach though so I'm open to changing it to that if there's a good reason.

> 2. Should the availability of the delete button be in the Profile ViewModel? Seems weird to make the MainPage responsible for it here. Esp. considering the future where dynamic profiles (and defaults) might need to be "hidden", not "deleted"

Hmm... I guess the design would be something like this...
- Profiles.cpp holds a list of in-box profile guids and dynamic profile guids
- bind DeleteButton.IsEnabled to a guid-check
- present a text block with why it's disabled based on the guid check

I'm sold! Changing it to this design.

zadjii-msft commented on src/cascadia/TerminalSettingsEditor/Profiles.xaml:
```diff
@@ -83,6 +83,52 @@ the MIT License. See LICENSE in the project root for license information. -->
+                            Style="{StaticResource BaseButtonStyle}"
+                            Margin="{StaticResource StandardControlSpacing}"
+                            IsEnabled="{x:Bind State.CanDeleteProfile}">
+                        <Button.Resources>
```
Should the styling for these buttons be in `CommonResources.xaml`? I'm _presuming_ they're styled the same?

carlos-zamora replied:
Creating a new `DeleteButtonStyle`. _Unfortunately_, I can't add the theming, font icon, and layout... I would have to retemplate it, I think? So really I just added the margin to the BaseButtonStyle hahaha.

zadjii-msft commented on src/cascadia/TerminalSettingsEditor/ColorSchemes.cpp:
```diff
@@ -120,6 +133,68 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
+
+        auto dialog{ FindName(L"RenameDialog").try_as<ContentDialog>() };
+        dialog.Title(winrt::box_value(title));
+        dialog.XamlRoot(sender.as<UIElement>().XamlRoot());
```
(see the `themingLambda` in `AppLogic.cpp` for the `RwquestedTheme` hack)

zadjii-msft commented on src/cascadia/TerminalSettingsEditor/ColorSchemes.cpp:
```diff
@@ -120,6 +133,68 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
+        return false;
+    }
+
+    void ColorSchemes::Delete_Click(IInspectable const& /*sender*/, RoutedEventArgs const& /*e*/)
```
Now this one, I like this one much better

zadjii-msft reviewed (changes requested):
Okay so I'm blocking over the questions I had, and to see how the `TextBox` + `Dialog` thing ends up working out

mdtauk commented:
> Okay so I'm blocking over the questions I had, and to see how the `TextBox` + `Dialog` thing ends up working out

https://github.com/microsoft/microsoft-ui-xaml/issues/3804

carlos-zamora commented:
> As a person who enters a blind rage (I kid) when he encounters a button that is disabled for what appears to be _no reason whatsoever_, I would very much like for us to have a tooltip or a disclaimer textblock in a slightly smaller font size explaining why a profile or colorscheme cannot be deleted.

Tried adding a tooltip, but [tooltips aren't supported on disabled buttons](https://github.com/microsoft/microsoft-ui-xaml/issues/1149). They can't get focus either. I've searched a little bit and found that a workaround may be to wrap it in a grid, place an invisible rectangle, and add a tooltip to _that_, but I'm concerned that this approach won't be accessible (particularly because of the focus issue). So I'm just going with the "disclaimer textblock" approach instead.

mdtauk commented:
> > As a person who enters a blind rage (I kid) when he encounters a button that is disabled for what appears to be _no reason whatsoever_, I would very much like for us to have a tooltip or a disclaimer textblock in a slightly smaller font size explaining why a profile or colorscheme cannot be deleted.
> 
> Tried adding a tooltip, but [tooltips aren't supported on disabled buttons](https://github.com/microsoft/microsoft-ui-xaml/issues/1149). They can't get focus either. I've searched a little bit and found that a workaround may be to wrap it in a grid, place an invisible rectangle, and add a tooltip to _that_, but I'm concerned that this approach won't be accessible (particularly because of the focus issue). So I'm just going with the "disclaimer textblock" approach instead.

I don't think it is a good idea to break expected behaviours - but adding some caption text saying something like (This profile is built in) etc to provide context for the reason why it cannot be deleted/removed.

carlos-zamora commented:
Spoke with Dustin and Kayla offline, to get around the ContentDialog issue, we'll do the following design: 

Upon clicking the rename button, the ComboBox will be replaced with a TextBox, and the rename button will be replaced with a confirm and a cancel button. The TextBox will be autopopulated with the selected color scheme's name. It'll look something like this:
![Mockup](https://user-images.githubusercontent.com/11050425/102421597-6bd50b80-3fca-11eb-8c2f-9a8455aeb231.png)
Accent color will be used for the accept button.

DHowett commented on src/cascadia/TerminalSettingsEditor/MainPage.cpp:
```diff
@@ -235,19 +235,12 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
+        state.DeleteProfile({ this, &MainPage::_DeleteProfile });
 
-        contentFrame().Navigate(xaml_typename<Editor::Profiles>(), *state);
+        contentFrame().Navigate(xaml_typename<Editor::Profiles>(), state);
```
not the hugest fan of having the _INCOMING STATE_ contain an event binding, but...

carlos-zamora replied:
Yeah. It does feel a bit weird to add the event binding to the state. But I also don't know what else we can do haha. Open to suggestions!

DHowett commented on src/cascadia/TerminalSettingsEditor/Profiles.cpp:
```diff
@@ -39,6 +39,37 @@ static const std::set<std::wstring_view> ProfileGeneratorNamespaces{
+            // in-box profile
+            return false;
+        }
+        else if (LegacyDynamicProfileGuids.find(guid) != LegacyDynamicProfileGuids.end() ||
+                 ProfileGeneratorNamespaces.find(source) != ProfileGeneratorNamespaces.end())
```
why not just "has a source"? That covers all of these cases without us having to hardcode a list of sources.

carlos-zamora replied:
Changing to "has a source". Curious, what's the point of the legacy dynamic profile guids again? I ran into them and thought they were important for a situation like this.

DHowett commented on src/cascadia/TerminalSettingsEditor/Profiles.cpp:
```diff
@@ -265,4 +296,9 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
         return _State.Profile().CursorShape() == TerminalControl::CursorStyle::Vintage;
     }
+
+    // -------------------------------- WinRT Events ---------------------------------
+    // Winrt events need a method for adding a callback to the event and removing the callback.
+    // These macros will define them both for you.
+    DEFINE_EVENT_WITH_TYPED_EVENT_HANDLER(ProfilePageNavigationState, DeleteProfile, _DeleteProfileHandlers, Editor::ProfilePageNavigationState, Editor::DeleteProfileEventArgs);
```
can we switch to the oneshot macro _please_

DHowett commented on src/cascadia/TerminalSettingsEditor/ColorSchemes.cpp:
```diff
@@ -89,6 +89,16 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
         CurrentColorScheme(colorScheme);
         _UpdateColorTable(colorScheme);
+
+        // Set the text disclaimer for the text box
+        hstring disclaimer{};
+        const std::wstring schemeName{ colorScheme.Name() };
+        if (InBoxSchemes.find(schemeName) != InBoxSchemes.end())
+        {
+            // load disclaimer for in-box profiles
+            disclaimer = RS_(L"ColorScheme_DeleteButtonDisclaimerInBox");
+        }
```
why is this not bound to a visibility like the Base Layer one is?

DHowett replied:
Or, why is this not bound to `CanDeleteCurrentScheme`?

carlos-zamora replied:
I could simplify this down to just toggling visibility, but I'm wondering if we'll have other disclaimers here. Particularly thinking about Proto-Extensions. We could just add an if-statement here to detect that a scheme came from a proto-extension and load a different disclaimer for it.

DHowett commented on src/cascadia/TerminalSettingsEditor/Profiles.cpp:
```diff
@@ -18,6 +18,25 @@ using namespace winrt::Windows::Storage::AccessCache;
+    { 0x0caa0dad, 0x35be, 0x5f56, { 0xa8, 0xff, 0xaf, 0xce, 0xee, 0xaa, 0x61, 0x01 } } // Command Prompt
+};
+
+static const std::set<winrt::guid> LegacyDynamicProfileGuids{
+    { 0x574e775e, 0x4f2a, 0x5b96, { 0xac, 0x1e, 0xa2, 0x96, 0x2a, 0x40, 0x23, 0x36 } }, // Powershell Core
+    { 0x58ad8b0c, 0x3ef8, 0x5f4d, { 0xbc, 0x6f, 0x13, 0xe4, 0xc0, 0x0f, 0x25, 0x30 } }, // Debian
+    { 0x2c4de342, 0x38b7, 0x51cf, { 0xb9, 0x40, 0x23, 0x09, 0xa0, 0x97, 0xf5, 0x18 } }, // Ubuntu
+    { 0x1777cdf0, 0xb2c4, 0x5a63, { 0xa2, 0x04, 0xeb, 0x60, 0xf3, 0x49, 0xea, 0x7c } }, // Alpine
+    { 0xc6eaf9f4, 0x32a7, 0x5fdc, { 0xb5, 0xcf, 0x06, 0x6e, 0x8a, 0x4b, 0x1e, 0x40 } } // Ubuntu-18.04
+};
```
no.

DHowett replied:
There's infinitely many of these, and they're all identifiable by the presence of a source property.

DHowett commented on src/cascadia/TerminalSettingsEditor/Profiles.cpp:
```diff
@@ -18,6 +18,25 @@ using namespace winrt::Windows::Storage::AccessCache;
+static const std::set<std::wstring_view> ProfileGeneratorNamespaces{
+    L"Windows.Terminal.Wsl",
+    L"Windows.Terminal.Azure",
+    L"Windows.Terminal.PowershellCore"
```
No.
There's infinitely many of these, and why would we hardcode a econd list of them?

DHowett commented on src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:
```diff
@@ -775,4 +772,19 @@
+    <value>Rename</value>
+  </data>
+  <data name="Profile_DeleteButtonDisclaimerInBox" xml:space="preserve">
+    <value>This profile cannot be deleted because it is shipped in-box.</value>
```
@cinnamon-msft to help with wording.

> This profile cannot be deleted because it is part of Terminal's default settings. You can hide it by visiting the Advanced page.

I really do almost think that if we're not letting people delete them we should make the button _effectively delete them_ by hiding.

carlos-zamora replied:
So, I have a few issues with the delete --> hide change:
- we already have a "Hide profile from dropdown" checkbox under the Advanced tab
- deleting and hiding are two completely different operations. Hiding removes it from the dropdown, but you can still select it as the default profile. Deleting completely removes the profile from the system.

I think it'd be better to...
- wait until Application State is implemented (#8324). Particularly the deleting dynamic profile part.
- keep the delete button as disabled for now. But when #8324 comes in, we enable it and update the Application State
- move "Hide" to the general tab and use [this approach](https://github.com/microsoft/terminal/pull/8403#issuecomment-744852588) to denote hidden items

cinnamon-msft replied:
I'm afraid of having the same button switch from "hide" and "delete" between different profiles. This may cause confusion and people may accidentally delete a profile they wanted to hide.

I like the idea of moving the "hide profile from dropdown" checkbox above the delete button so at least people can find it more easily.

As for wording, I'm thinking: "This profile cannot be deleted because it is included by default."

DHowett commented on src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:
```diff
@@ -775,4 +772,19 @@
+    <value>This profile cannot be deleted because it is shipped in-box.</value>
+  </data>
+  <data name="Profile_DeleteButtonDisclaimerDynamic" xml:space="preserve">
+    <value>This profile cannot be deleted because it is a dynamic profile.</value>
```
Wording

htcfreek replied:
I think maybe "auto-generated" could be a better phrase here.

cinnamon-msft replied:
"This profile cannot be deleted because it is automatically generated." ?

DHowett commented on src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:
```diff
@@ -775,4 +772,19 @@
+    <value>This profile cannot be deleted because it is a dynamic profile.</value>
+  </data>
+  <data name="ColorScheme_DeleteButtonDisclaimerInBox" xml:space="preserve">
+    <value>This color scheme cannot be deleted or renamed because it is shipped in-box.</value>
```
once again "shipping" and "in-box" don't mean things to like 99% of our userbase

htcfreek replied:
Can you use the phrase "it's preinstalled with Windows Terminal" or can we simply hide the buttons on this schemes/profiles?

cinnamon-msft replied:
"This profile cannot be deleted because it is included by default." ?

carlos-zamora commented:
# Demo 12/17/2020

Now with gifs!

## Color Schemes
![Color Schemes Demo](https://user-images.githubusercontent.com/11050425/102513812-4dfeb980-4051-11eb-813a-57fe7d12d757.gif)

## Profiles
![Profiles Demo](https://user-images.githubusercontent.com/11050425/102513986-8a321a00-4051-11eb-9596-ca86b4be98f3.gif)

DHowett commented:
I’m still holding out for that flyout on the delete button asking the user to confirm it! It’s a destructive operation, doing it instantly without asking will result in data loss

htcfreek reviewed (commented):
Text suggestion.

carlos-zamora commented:
> I’m still holding out for that flyout on the delete button asking the user to confirm it! It’s a destructive operation, doing it instantly without asking will result in data loss

![Color Scheme Confirmation Dialog](https://user-images.githubusercontent.com/11050425/102523758-99b76000-405d-11eb-9bc1-cc6790cc5f19.png)

![Profile Confirmation Dialog](https://user-images.githubusercontent.com/11050425/102523788-a63bb880-405d-11eb-903d-ee98c550d002.png)

Since you can see the name behind the dialog, figured it'd be fine to show not include the name of the profile/scheme in the content dialog. But I'm open to changing that. Just let me know 😊

DHowett commented:
Can we not use a simple flyout? Content dialog seems too heavyhanded. I don’t mean “flyout”, i mean `Flyout`. The class.

htcfreek commented:
> > I’m still holding out for that flyout on the delete button asking the user to confirm it! It’s a destructive operation, doing it instantly without asking will result in data loss
> 
> ![Color Scheme Confirmation Dialog](https://user-images.githubusercontent.com/11050425/102523758-99b76000-405d-11eb-9bc1-cc6790cc5f19.png)
> 
> ![Profile Confirmation Dialog](https://user-images.githubusercontent.com/11050425/102523788-a63bb880-405d-11eb-903d-ee98c550d002.png)
> 
> Since you can see the name behind the dialog, figured it'd be fine to show not include the name of the profile/scheme in the content dialog. But I'm open to changing that. Just let me know 😊

Should we define the "no" button as default?

carlos-zamora commented:
Replaced the contentdialog with the flyout so it feels less aggressive. Loving it!

## Demo - Flyouts
![Delete Profile](https://user-images.githubusercontent.com/11050425/102535115-5e249200-406d-11eb-9410-336e0f6f655c.png)

![Delete Color Scheme](https://user-images.githubusercontent.com/11050425/102535156-6ed50800-406d-11eb-94a8-7f4a99f968f2.png)

DHowett commented on src/cascadia/TerminalSettingsModel/Profile.idl:
```diff
@@ -18,8 +18,7 @@ namespace Microsoft.Terminal.Settings.Model
 
-    [flags]
-    enum ConvergedAlignment {
+    [flags] enum ConvergedAlignment {
```
this was a holdover from before your merge. revert it.

DHowett commented on src/cascadia/TerminalSettingsModel/Microsoft.Terminal.Settings.ModelLib.vcxproj.filters:
```diff
@@ -89,4 +89,4 @@
     </Filter>
   </ItemGroup>
-</Project>
+</Project>
```
this was a holdover from before your merge. revert it.

DHowett commented on src/cascadia/TerminalSettingsModel/Microsoft.Terminal.Settings.ModelLib.vcxproj:
```diff
@@ -214,4 +214,4 @@
 
   <Import Project="$(SolutionDir)build\rules\CollectWildcardResources.targets" />
-</Project>
+</Project>
```
this was a holdover from before your merge. revert it.

DHowett commented on src/cascadia/TerminalSettingsModel/CascadiaSettings.cpp:
```diff
@@ -225,7 +225,7 @@ winrt::Microsoft::Terminal::Settings::Model::Profile CascadiaSettings::CreateNew
 
     // Give the new profile a distinct name so a guid is properly generated
-    const winrt::hstring newName{ fmt::format(L"Profile {}", _allProfiles.Size()) };
+    const hstring newName{ fmt::format(L"Profile {}", _allProfiles.Size()) };
```
why?

DHowett commented on src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:
```diff
@@ -741,4 +762,40 @@
+    <value>Are you sure you want to delete this color scheme?</value>
+  </data>
+  <data name="ColorScheme_RenameAccept.[using:Windows.UI.Xaml.Controls]ToolTipService.ToolTip" xml:space="preserve">
+    <value>Accept rename</value>
```
We're going to have multiple versions of these, so I'd call it `Accept` and `Cancel` and make them use a common `Uid` (not one that's namespaced in ColorScheme)`

DHowett commented on src/cascadia/TerminalSettingsEditor/Profiles.idl:
```diff
@@ -50,12 +50,27 @@ namespace Microsoft.Terminal.Settings.Editor
 
         ProfileViewModel Profile { get; };
+
+        Boolean CanDeleteProfile { get; };
+        Boolean IsBaseLayer;
```
technically this is viewmodel stuff

DHowett commented on src/cascadia/TerminalSettingsEditor/Profiles.idl:
```diff
@@ -50,12 +50,27 @@ namespace Microsoft.Terminal.Settings.Editor
+
+        event Windows.Foundation.TypedEventHandler<ProfilePageNavigationState, DeleteProfileEventArgs> DeleteProfile;
+
+        // This needs to be the theme we are currently using, not the one we're changing to.
+        // We need the theme to be able to force the ContentDialog to follow the appropriate
+        // theme chosen by the user. See Delete_Click() for more details.
+        Windows.UI.Xaml.ElementTheme Theme;
```
we've removed all dialogs, so we don't need this any more?

DHowett commented on src/cascadia/TerminalSettingsEditor/Profiles.h:
```diff
@@ -60,24 +61,42 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
 
     private:
         Windows::Foundation::Collections::IMapView<hstring, Model::ColorScheme> _Schemes;
+        std::function<void(Editor::ProfileViewModel&)> _pfnDeleteProfile{ nullptr };
```
we have a pfnDeleteProfile _and_ a typed event?

DHowett commented on src/cascadia/TerminalSettingsEditor/Profiles.h:
```diff
@@ -60,24 +61,42 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
+        DeleteProfileEventArgs(guid profileGuid) :
+            _ProfileGuid(profileGuid) {}
+
+        GETSET_PROPERTY(guid, ProfileGuid);
```
this only needs to be a getter, and the idl only needs the `get;` decoration

DHowett commented on src/cascadia/TerminalSettingsEditor/Profiles.cpp:
```diff
@@ -64,6 +100,24 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
         }
+
+        // Set the text disclaimer for the text box
+        // For an unknown reason, data-binding the text does not compile (using x:Bind)
```
every compiler error has a reason.

DHowett replied:
(don't leave comments like this.)

carlos-zamora replied:
Removing the "for an unknown reason..." part. IIRC, the error came up in the XAML code, but I don't have XAML symbols with me so I can't look into it all the way.

DHowett commented on src/cascadia/TerminalSettingsEditor/Profiles.cpp:
```diff
@@ -18,8 +19,43 @@ using namespace winrt::Windows::Storage::AccessCache;
 using namespace winrt::Windows::Storage::Pickers;
 using namespace winrt::Microsoft::Terminal::Settings::Model;
 
+static const std::set<winrt::guid> InBoxProfileGuids{
```
please try to avoid statically initialized stl types. just like you don't have a `static hstring`, don't have a `static set`.

Use a constexpr std::array or something.

DHowett reviewed (changes requested):
14/30

zadjii-msft commented on src/cascadia/TerminalSettingsEditor/MainPage.cpp:
```diff
@@ -233,19 +235,12 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
     void MainPage::_Navigate(const Editor::ProfileViewModel& profile)
     {
-        auto state{ winrt::make_self<ProfilePageNavigationState>(profile, _settingsClone.GlobalSettings().ColorSchemes(), *this) };
+        auto state{ winrt::make<ProfilePageNavigationState>(profile, _settingsClone.GlobalSettings().ColorSchemes(), *this) };
```
woah wait why is this a `make` and not a `make_self`? Just curious

carlos-zamora replied:
No reason, really. I've been using `winrt::make` for all the other navigation states so I guess to be consistent

zadjii-msft reviewed (approved):
Okay yea, passing the event handler in the state is wack, but I don't have a better idea.

DHowett commented on src/cascadia/TerminalSettingsModel/ColorScheme.cpp:
```diff
@@ -42,27 +42,24 @@ static constexpr std::array<std::string_view, 16> TableColors = {
+    _Foreground{ defaultFg },
+    _Background{ defaultBg },
+    _SelectionBackground{ DEFAULT_FOREGROUND },
+    _CursorColor{ cursorColor }
```
LOVE IT

DHowett reviewed (approved):
Looks great. I love it.

stephenanthonylewis reviewed (approved):
L

DHowett commented:
@stephenanthonylewis Do explain why it was necessary to sign off on an already-completed pull request from three months ago _three times_ and e-mail the entire team about it :wink:
